plasmapy/particles/tests/test_particle_class.py

```diff
@@ -997,3 +998,110 @@ def test_customized_particle_repr(cls, kwargs, expected_repr):
             f"from_str: {from_str}"
             f"from_repr: {from_repr}"
         )
+
+
+custom_particles_from_json_tests = [
+    (
+        DimensionlessParticle,
+        {"mass": 5.2, "charge": 6.3},
+        '{"plasmapy_particle": {"type": "DimensionlessParticle", "module": "plasmapy.particles.particle_class", "date_created": "2020-06-08", "init_args": {"mass": "5.2", "charge": "6.3"}}}'
+    ),
+    (
+        DimensionlessParticle,
+        {"mass": 5.2},
+        '{"plasmapy_particle": {"type": "DimensionlessParticle", "module": "plasmapy.particles.particle_class", "date_created": "2020-06-08", "init_args": {"mass": "5.2", "charge": "nan"}}}'
+    ),
+    (
+        CustomParticle,
+        {"mass": 5.12 * u.kg, "charge": 6.2 * u.C},
+        '{"plasmapy_particle": {"type": "CustomParticle", "module": "plasmapy.particles.particle_class", "date_created": "2020-06-08", "init_args": {"mass": "5.12 kg", "charge": "6.2 C"}}}'
+    ),
+    (
+        CustomParticle,
+        {"mass": 5.12 * u.kg},
+        '{"plasmapy_particle": {"type": "CustomParticle", "module": "plasmapy.particles.particle_class", "date_created": "2020-06-08", "init_args": {"mass": "5.12 kg", "charge": "nan C"}}}'
+    )
+]
+
+
+@pytest.mark.parametrize("cls, kwargs, json_string", custom_particles_from_json_tests)
+def test_custom_particles_from_json(cls, kwargs, json_string):
+    """Test the attributes of dimensionless and custom particles generated from
+    JSON representation"""
+    instance = cls(**kwargs)
+    instance_from_json = from_json(json_string)
+    assert u.isclose(instance.mass, instance_from_json.mass, equal_nan=True), pytest.fail(
+            f"Expected a mass value of {instance.mass}\n"
+            f"Received a mass value of {instance_from_json.mass}"
+        )
+    assert u.isclose(instance.charge, instance_from_json.charge, equal_nan=True), pytest.fail(
+            f"Expected a charge value of {instance.charge}\n"
+            f"Received a charge value of {instance_from_json.charge}"
+        )
```

Comment:
Could be good to also compare `instance == instance_from_json` here, just like you do below!

Reply:
Thank you for the suggestion. I will look into adding that check as well.

Reply:
Interestingly, I tried this check and it lead to some interesting behaviour:
```python
>>> c = CustomParticle(mass = 5.2 * u.kg, charge = 5.2 * u.C)
>>> c1 = from_json('{"plasmapy_particle": {"type": "CustomParticle", "module": "plasmapy.particles.particle_class", "date_created": "2020-06-08", "init_args": {"mass": "5.2 kg", "charge": "5.2 C"}}}')
>>> c == c1
False
>>> c.mass == c1.mass
True
>>> c.charge == c1.charge
True
```
Any thoughts?

Reply:
Oh yeah, floating point comparisons, especially when compared to the truncated version saved in JSON? That's a bummer, but I guess it is what it is...

Reply:
Oh wait, I scrolled to the right. Still, I'd expect this to be a floating point thing. I can run a debugger to make sure.

Reply:
Just a thought, but could it be because there is no `__eq__` defined for the `CustomParticle` class? I noticed that `__eq__` is defined for `Particle` and this seems to pass the test. I'm guessing that it may be using the default behaviour for equality of `CustomParticle`.

Reply:
Welp. I just realized `Particle` has `__eq__`, but `CustomParticle` doesn't! Definitely out of scope for this issue :laughing: